diff --git a/tools/console/daemon/io.c b/tools/console/daemon/io.c
--- a/tools/console/daemon/io.c	2023-04-23 14:01:40.340769254 +0800
+++ b/tools/console/daemon/io.c	2023-04-23 13:56:38.932171183 +0800
@@ -56,9 +56,9 @@
 #define MAX_STRLEN(x) ((sizeof(x) * CHAR_BIT + CHAR_BIT-1) / 10 * 3 + 2)
 
 /* How many events are allowed in each time period */
-#define RATE_LIMIT_ALLOWANCE 30
+#define RATE_LIMIT_ALLOWANCE 300000 ///30
 /* Duration of each time period in ms */
-#define RATE_LIMIT_PERIOD 200
+#define RATE_LIMIT_PERIOD 200 ///200
 
 extern int log_reload;
 extern int log_guest;

